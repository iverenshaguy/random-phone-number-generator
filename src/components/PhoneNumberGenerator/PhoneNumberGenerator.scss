.rc-PhoneNumberGenerator {
  .main {
    display: flex;
    flex-direction: column;
    padding: 20px;
  }

  .select {
    border-radius: 0;
  }

  table {
   td, th {
     font-size: 1rem;
     padding: 0.2rem;
   }
  }

  .result-card {
    min-width: 600px;
    max-width: 912px;
    min-height: 416px;
    border: 1px solid #ced4da;

    .default-message {
      display: flex;
      flex-direction: column;
      justify-content: center;
      align-items: center;

      p {
        font-size: 1.4em;
      }
    }
  }

  @media screen and (max-width: 992px) {
    .result-card {
      min-width: 330px;
      min-height: 416px;
      border: 1px solid #ced4da;
    }
  }
}
