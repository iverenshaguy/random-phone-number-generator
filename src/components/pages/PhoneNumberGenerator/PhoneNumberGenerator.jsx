import React, { Component } from 'react';
import Loader from 'react-loader-spinner'
import { Button, InputGroup, InputGroupAddon, Input, Table } from 'reactstrap';
import CsvDownloader from 'react-csv-downloader';

import Pagination from '../../shared/Pagination/Pagination';
import Modal from '../../shared/Modal/Modal';
import generateUniqueNumbers from '../../../utils/generateUniqueNumbers';
import sortNumbers from '../../../utils/sortNumbers';

import './PhoneNumberGenerator.scss';


export default class PhoneNumberGenerator extends Component {
  state = {
    count: 1000,
    sort: 'ASC',
    loading: false,
    previousResults: [],
    results: [],
    metadata: {
      page: 1,
      perPage: 10,
      totalCount: 0,
    }
  }

  handlePageChange = (page) => {
    this.setState({
      metadata: {
        ...this.state.metadata,
        page
      }
    });
  }

  handleChange = ({ target: { name, value } }) => {
    const prevSort = this.state.sort;

    this.setState({
      [name]: value
    }, () => {
      const { sort, results } = this.state;

      if (prevSort !== sort && results.length) {
        this.sortNumbers();
      }
    });
  }

  sortNumbers = () => {
    const sorted = sortNumbers(this.state.results, this.state.sort);

    this.updateResults(sorted);
  }

  generateNumbers = () => {
    this.setState({
      loading: true
    }, () => {
      const { previousResults, count, sort } = this.state;
      const results = generateUniqueNumbers(previousResults, count, sort);

      this.updateResults(results);
    });
  }

  updateResults = (results) => {
    this.setState({
      results,
      loading: false,
      previousResults: results,
      metadata: {
        ...this.state.metadata,
        page: 1,
        totalCount: results.length,
      }
    });
  }

  renderDefaultMessage = () => (
    <div className="default-message">
      <p className="px-5">Generate some numbers by supplying count of less than or equal to 10000 and sort of ascending or descending order</p>
    </div>
  );


  renderResult = () => {
    const { results, loading, metadata: { perPage, page } } = this.state;
    const offset = (page - 1) * perPage;
    const resultsToRender = results.slice(offset, offset + perPage);

    if (loading) {
      return (
        <div className="my-auto">
          <Loader type="Ball-Triangle" color="#ced4da" height={80} width={80} />
        </div>
      );
    }


    return (
      <Table className="w-100">
        <thead>
          <tr>
            <th>#</th>
            <th>Number</th>
          </tr>
        </thead>
        <tbody>
          {resultsToRender.map((result, index) => (<tr key={result}>
            <td>{index + offset + 1}</td>
            <td>{result}</td>
          </tr>))}
        </tbody>
      </Table>
    );
  }

  renderModalActionButton= () => {
    const datas = this.state.results.map((number, index) => ({
      id: index + 1,
      number: number.toString()
    }));

    const columns = [{
      id: 'id',
      displayName: 'ID'
    }, {
      id: 'number',
      displayName: 'Numbers'
    }];

    console.log(datas);


    return (
      <CsvDownloader datas={datas} columns={columns} filename="phone-numbers">
        <Button color="secondary" onClick={this.saveResults}>Save Results</Button>
      </CsvDownloader>
    );
  }

  renderModalBody = () => {
    const { sort, results } = this.state;
    const length = results.length;
    const minimum = (!sort || sort === 'ASC') ? results[0] : results[length - 1];
    const maximum = (!sort || sort === 'ASC') ? results[length - 1] : results[0];
    const order = sort === 'ASC' ? 'Ascending' : 'Descending';

    return (
      <dl className="row mx-auto mb-0 px-5">
        <dt className="pl-5 col-sm-4">Order:</dt>
        <dd className="pl-5 col-sm-8">{order}</dd>
        <dt className="pl-5 col-sm-4">Minimum:</dt>
        <dd className="pl-5 col-sm-8">{minimum}</dd>
        <dt className="pl-5 col-sm-4">Maximum:</dt>
        <dd className="pl-5 col-sm-8">{maximum}</dd>
        <dt className="pl-5 col-sm-4">Count:</dt>
        <dd className="pl-5 col-sm-8">{length}</dd>
      </dl>
    );
  }

  render() {
    const { metadata, results, count, sort, loading } = this.state;
    const disableButton = count > 10000 || count < 1;
    const ungenerated = !results.length && !loading;

    return (
      <div className="container-fluid rc-PhoneNumberGenerator mt-5">
        <div className="row justify-content-center main">
          <h4 className="text-center m-3 m-3-md m-5-l">Random Phone Number Generator</h4>
          <div className="d-flex justify-content-between">
            <Input className="w-50 mr-2" type="select" name="sort" value={sort} onChange={this.handleChange}>
              <option>ASC</option>
              <option>DESC</option>
            </Input>
            <InputGroup className="w-100 mb-3">
              <Input type="number" placeholder="Numbers Count" name="count" value={count} onChange={this.handleChange} />
              <InputGroupAddon addonType="append">
                <Button className="w-100" disabled={disableButton} onClick={this.generateNumbers}>Generate</Button>
              </InputGroupAddon>
            </InputGroup>
          </div>
          <Modal
            disabled={!results.length}
            title="Results"
            label="View/Save Results"
            actionButton={this.renderModalActionButton()}
          >
            {this.renderModalBody()}
          </Modal>
          <div className="card result-card text-center mb-3 pt-3">
            <div className="card-body d-flex justify-content-center result-content">
              {ungenerated ? this.renderDefaultMessage() : this.renderResult()}
            </div>
          </div>
          <Pagination metadata={metadata} onPageChange={this.handlePageChange} />
        </div>
      </div>
    );
  }
}
