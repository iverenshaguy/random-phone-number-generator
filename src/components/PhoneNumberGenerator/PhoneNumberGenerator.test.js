import React from 'react';
import PropTypes from 'prop-types';

import PhoneNumberGenerator from './PhoneNumberGenerator';

describe('PhoneNumberGenerator', () => {
  const props = {
    onPageChange: jest.fn(),
    metadata: {
      page: 1,
      perPage: 10,
      totalCount: 450,
    }
  };
  const wrapper = mount(
    <PhoneNumberGenerator {...props} />, {
      context: {
        windowWidth: 1000,
        windowHeight: 1500,
        isMobileSized: false,
      },
      childContextTypes: {
        windowWidth: PropTypes.number,
        windowHeight: PropTypes.number,
        isMobileSized: PropTypes.bool,
      }
    }
  );

  it('should render PhoneNumberGenerator component correctly', () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should handle input box change', () => {
    expect(wrapper.find('input[name="count"]').instance().value).toEqual("1000");

    wrapper.find('input[name="count"]').simulate('change', { target: { name: 'count', value: 2000 } });
    expect(wrapper.find('input[name="count"]').instance().value).toEqual("2000");
  });

  it('should handle select box change', () => {
    expect(wrapper.find('select[name="sort"]').instance().value).toEqual('ASC');

    wrapper.find('select[name="sort"]').simulate('change', { target: { name: 'sort', value: 'DESC' } });
    expect(wrapper.find('select[name="sort"]').instance().value).toEqual("DESC");
  });

  it('should generate unique numbers and hide default message on generate button click', () => {
    expect(wrapper.find('.default-message').length).toEqual(1);
    wrapper.find('.input-group-append button').simulate('click');

    expect(wrapper.find('.default-message').length).toEqual(0);
  });

  it('should sort generated numbers on when select button is changed', () => {
    const sortNumbersSpy = jest.spyOn(wrapper.instance(), 'sortNumbers');
    wrapper.find('select[name="sort"]').simulate('change', { target: { name: 'sort', value: 'ASC' } });

    expect(sortNumbersSpy).toHaveBeenCalled();
  });

  it('should change page in state when a pagination item is clicked', () => {
    expect(wrapper.state().metadata.page).toEqual(1);
    wrapper.find('.page-item a').at(4).simulate('click');

    expect(wrapper.state().metadata.page).toEqual(3);
  });
});