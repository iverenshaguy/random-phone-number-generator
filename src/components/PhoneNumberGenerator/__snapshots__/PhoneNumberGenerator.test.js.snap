// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PhoneNumberGenerator should render PhoneNumberGenerator component correctly 1`] = `
<PhoneNumberGenerator
  metadata={
    Object {
      "page": 1,
      "perPage": 10,
      "totalCount": 450,
    }
  }
  onPageChange={[MockFunction]}
>
  <div
    className="container-fluid rc-PhoneNumberGenerator mt-5"
  >
    <div
      className="row justify-content-center main"
    >
      <h4
        className="text-center m-3 m-3-md m-5-l"
      >
        Random Phone Number Generator
      </h4>
      <div
        className="d-flex justify-content-between"
      >
        <Input
          className="w-50 mr-2"
          name="sort"
          onChange={[Function]}
          type="select"
          value="ASC"
        >
          <select
            className="w-50 mr-2 form-control"
            name="sort"
            onChange={[Function]}
            value="ASC"
          >
            <option>
              ASC
            </option>
            <option>
              DESC
            </option>
          </select>
        </Input>
        <InputGroup
          className="w-100 mb-3"
          tag="div"
        >
          <div
            className="w-100 mb-3 input-group"
          >
            <Input
              name="count"
              onChange={[Function]}
              placeholder="Numbers Count"
              type="number"
              value={1000}
            >
              <input
                className="form-control"
                name="count"
                onChange={[Function]}
                placeholder="Numbers Count"
                type="number"
                value={1000}
              />
            </Input>
            <InputGroupAddon
              addonType="append"
              tag="div"
            >
              <div
                className="input-group-append"
              >
                <Button
                  className="w-100"
                  color="secondary"
                  disabled={false}
                  onClick={[Function]}
                  tag="button"
                >
                  <button
                    aria-label={null}
                    className="w-100 btn btn-secondary"
                    disabled={false}
                    onClick={[Function]}
                    type="button"
                  >
                    Generate
                  </button>
                </Button>
              </div>
            </InputGroupAddon>
          </div>
        </InputGroup>
      </div>
      <Modal
        actionButton={
          <CsvDownload
            bom={true}
            columns={
              Array [
                Object {
                  "displayName": "ID",
                  "id": "id",
                },
                Object {
                  "displayName": "Numbers",
                  "id": "number",
                },
              ]
            }
            datas={Array []}
            filename="phone-numbers"
            noHeader={false}
            separator=","
          >
            <Button
              color="secondary"
              tag="button"
            >
              Save Results
            </Button>
          </CsvDownload>
        }
        disabled={true}
        label="View/Save Results"
        title="Results"
      >
        <Button
          className="w-50 h-25 mx-auto mb-4"
          color="secondary"
          disabled={true}
          onClick={[Function]}
          tag="button"
        >
          <button
            aria-label={null}
            className="w-50 h-25 mx-auto mb-4 btn btn-secondary disabled"
            disabled={true}
            onClick={[Function]}
            type="button"
          >
            View/Save Results
          </button>
        </Button>
        <Modal
          autoFocus={true}
          backdrop={true}
          backdropTransition={
            Object {
              "mountOnEnter": true,
              "timeout": 150,
            }
          }
          centered={false}
          fade={true}
          isOpen={false}
          keyboard={true}
          modalTransition={
            Object {
              "timeout": 300,
            }
          }
          onClosed={[Function]}
          onOpened={[Function]}
          returnFocusAfterClose={true}
          role="dialog"
          scrollable={false}
          toggle={[Function]}
          unmountOnClose={true}
          zIndex={1050}
        />
      </Modal>
      <div
        className="card result-card text-center mb-3 pt-3"
      >
        <div
          className="card-body d-flex justify-content-center result-content"
        >
          <div
            className="default-message"
          >
            <p
              className="px-5"
            >
              Generate some numbers by supplying count of less than or equal to 10000 and sort of ascending or descending order
            </p>
          </div>
        </div>
      </div>
      <Pagination
        metadata={
          Object {
            "page": 1,
            "perPage": 10,
            "totalCount": 0,
          }
        }
        onPageChange={[Function]}
      >
        <div
          className="mx-auto invisible"
        >
          <a
            activePage={1}
            firstPageText="«"
            getPageUrl={[Function]}
            hideFirstLastPages={false}
            innerClass="pagination"
            itemClass="page-item"
            itemsCountPerPage={10}
            lastPageText="»"
            nextPageText="⟩"
            onChange={[Function]}
            pageRangeDisplayed={10}
            prevPageText="⟨"
            totalItemsCount={0}
          >
            <ul
              className="pagination"
            >
              <t
                activeClass="active"
                disabledClass="disabled"
                href="#"
                isActive={false}
                isDisabled={true}
                itemClass="page-item"
                key="first"
                linkClass=""
                onClick={[Function]}
                pageNumber={1}
                pageText="«"
              >
                <li
                  className="page-item disabled"
                  onClick={[Function]}
                >
                  <a
                    className=""
                    href="#"
                  >
                    «
                  </a>
                </li>
              </t>
              <t
                activeClass="active"
                disabledClass="disabled"
                href="#"
                isActive={false}
                isDisabled={true}
                itemClass="page-item"
                key="prev-1"
                linkClass=""
                onClick={[Function]}
                pageNumber={-1}
                pageText="⟨"
              >
                <li
                  className="page-item disabled"
                  onClick={[Function]}
                >
                  <a
                    className=""
                    href="#"
                  >
                    ⟨
                  </a>
                </li>
              </t>
              <t
                activeClass="active"
                disabledClass="disabled"
                href="#"
                isActive={false}
                isDisabled={true}
                itemClass="page-item"
                key="next1"
                linkClass=""
                onClick={[Function]}
                pageNumber={1}
                pageText="⟩"
              >
                <li
                  className="page-item disabled"
                  onClick={[Function]}
                >
                  <a
                    className=""
                    href="#"
                  >
                    ⟩
                  </a>
                </li>
              </t>
              <t
                activeClass="active"
                disabledClass="disabled"
                href="#"
                isActive={false}
                isDisabled={true}
                itemClass="page-item"
                key="last"
                linkClass=""
                onClick={[Function]}
                pageNumber={0}
                pageText="»"
              >
                <li
                  className="page-item disabled"
                  onClick={[Function]}
                >
                  <a
                    className=""
                    href="#"
                  >
                    »
                  </a>
                </li>
              </t>
            </ul>
          </a>
        </div>
      </Pagination>
    </div>
  </div>
</PhoneNumberGenerator>
`;
